### PR TITLE
Feature/112 client connection retry

### PR DESCRIPTION
**Pull request by adenh93**

### CODE REVIEW

File: client/examples/publish_bincode.rs [hunk 1]
```diff
@@ -4,7 +4,7 @@ use selium::prelude::*;
 use selium::std::codecs::BincodeCodec;
 use serde::{Deserialize, Serialize};
 
-#[derive(Debug, Serialize, Deserialize)]
+#[derive(Clone, Debug, Serialize, Deserialize)]
```
Comment (petehayes102) [inline client/examples/publish_bincode.rs:7]: What's this `Clone` for?

File: client/src/connection.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+async fn connect_to_endpoint(addr: SocketAddr, config: ClientConfig) -> Result<Connection> {
+    let endpoint_addr = ENDPOINT_ADDRESS
+        .parse::<SocketAddr>()
+        .map_err(SeliumError::ParseEndpointAddressError)?;
+
+    let mut endpoint = Endpoint::client(endpoint_addr)?;
+    endpoint.set_default_client_config(config);
+    let connection = endpoint.connect(addr, "localhost")?.await?;
```
Comment (petehayes102) [inline client/src/connection.rs:98]: Seems a security risk hardcoding this to "localhost" given what the docs say, right?

> `server_name` must be covered by the certificate presented by the server. This prevents a connection from being intercepted by an attacker with a valid certificate for some other server.

File: client/src/crypto/cert.rs [hunk 3]
```diff
@@ -1,25 +1,29 @@
...
+use selium_std::errors::{Result, SeliumError};
 use std::{fs, path::Path};
 
 pub type KeyPair = (Vec<Certificate>, PrivateKey);
 
 fn load_key<T: AsRef<Path>>(path: T) -> Result<PrivateKey> {
     let path = path.as_ref();
-    let key = fs::read(path).context("failed to read private key")?;
+    let key =
+        fs::read(path).map_err(|_| SeliumError::InvalidKeys("failed to read private key."))?;
```
Comment (petehayes102) [inline client/src/crypto/cert.rs:11]: Did you want to attach the upstream error to your error type? That way you have causation. E.g. was it a permissions issue, a network mount gone bad etc.

> The Error trait’s source() method is implemented to return whichever field has a #[source] attribute or is named source, if any. This is for identifying the underlying lower level error that caused your error.

```
#[derive(Error, Debug)]
pub struct MyError {
    msg: String,
    #[source]  // optional if field name is `source`
    source: anyhow::Error,
}
```

File: client/src/keep_alive/backoff_strategy.rs [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+pub struct BackoffStrategy {
+    strategy_type: Strategy,
+    state: BackoffStrategyState,
+}
+
+impl BackoffStrategy {
+    fn new(strategy_type: Strategy) -> Self {
+        Self {
+            strategy_type,
+            ..Self::default()
```
Comment (petehayes102) [inline client/src/keep_alive/backoff_strategy.rs:46]: Very sleek ;)

File: protocol/Cargo.toml [hunk 5]
```diff
@@ -20,3 +20,4 @@ futures = "0.3"
 quinn = "0.10"
 serde = { version = "1.0", features = ["derive"] }
 tokio-util = { version = "0.7", features = ["codec"] }
+selium-std = { version = "0.1", path = "../standard" }
```
Comment (petehayes102) [inline protocol/Cargo.toml:23]: This is reaaaally picky but could you sort this alphabetically preesh? 😅

File: protocol/src/bistream.rs [hunk 6]
```diff
@@ -53,21 +53,21 @@ impl From<(SendStream, RecvStream)> for BiStream {
 }
 
 impl Sink<Frame> for BiStream {
-    type Error = anyhow::Error;
+    type Error = SeliumError;
 
-    fn poll_ready(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
```
Comment (petehayes102) [inline protocol/src/bistream.rs:58]: Is it bad that we define the error type on the `BiStream` and then completely ignore it? I get that the result is the same but seems like a future foot gun...? 🤷‍♂️

File: protocol/src/frame.rs [hunk 7]
```diff
@@ -17,7 +17,7 @@ pub enum Frame {
 }
 
 impl Frame {
-    pub fn get_length(&self) -> Result<u64> {
+    pub fn get_length(&self) -> Result<u64, SeliumError> {
```
Comment (petehayes102) [inline protocol/src/frame.rs:20]: Use `selium_std::errors::Result` instead here?

File: standard/Cargo.toml [hunk 8]
```diff
@@ -13,7 +13,9 @@ brotli = { version = "3.4", optional = true }
 bytes = "1.4"
 flate2 = { version = "1.0", features = ["zlib"], optional = true }
 lz4_flex = { version = "0.11", optional = true }
+quinn = "0.10"
 serde = { version = "1.0", optional = true }
+thiserror = "1.0"
```
Comment (petehayes102) [inline standard/Cargo.toml:18]: Should we delete `anyhow` dependency?

**Review by petehayes102 — CHANGES_REQUESTED** [review]
A slick PR as always mate! Just a few small bits and bobs. Happy for you to push back as/where you feel appropriate 😄

File: client/examples/publish_bincode.rs (hunk 1 above)
Comment (adenh93) [reply client/examples/publish_bincode.rs]: This was a mistake made in the generic constraints for the `KeepAlive` Sink impl 🤦‍♂️ 

Fixed!

File: protocol/src/bistream.rs (hunk 6 above)
Comment (adenh93) [reply protocol/src/bistream.rs:58]: Updated the `Sink` impls now 👌

**Review by petehayes102 — APPROVED** [review]

**Pull request merged**